File: src/Jenkins/JenkinsEndpoint.ts
import { btoa } from "../b64.js";
import { default as fetch, Headers } from "node-fetch";
import type { Endpoint } from "../Endpoint.js";
import type { Job } from "./Job.js";
import type { BuildDetail } from "./BuildDetail.js";
import type { RequestInit } from "node-fetch";
import type { endpointConfig } from "../config.js";

export class JenkinsEndpoint implements Endpoint {
  private readonly apiUrlSuffix = "api/json?pretty=false";
  private readonly headers: RequestInit;
  private readonly conf: endpointConfig;

  constructor(config: endpointConfig) {
    // Check env
    if (!process.env["JENKINS_TOKEN"]) {
      throw new Error("Envoirement variable JENKINS_TOKEN is not set");
    }
    if (!process.env["JENKINS_USER"]) {
      throw new Error("Envoirement variable JENKINS_USER is not set");
    }

    this.conf = config;

    const headers = new Headers();
    headers.set("Accept", "application/json");
    headers.set(
      "Authorization",
      "Basic " +
        btoa(process.env["JENKINS_USER"] + ":" + process.env["JENKINS_TOKEN"])
    );

    this.headers = {
      headers,
    };
  }

  public getLastFailedTimestamp(project: string) {
    return new Promise<number>((res, rej) => {
      fetch(
        `${this.conf.url}/job/${project}/${this.apiUrlSuffix}` +
          "&tree=lastFailedBuild[timestamp,url]",
        this.headers
      )
        .then((jobResp) => {
          (jobResp.json() as Promise<Job>)
            .then((job) => {
              if (job.lastFailedBuild) {
                if (job.lastFailedBuild.timestamp) {
                  res(job.lastFailedBuild.timestamp);
                } else if (job.lastFailedBuild.url) {
                  fetch(
                    job.lastFailedBuild.url + this.apiUrlSuffix,
                    this.headers
                  )
                    .then((buildResp) => {
                      (buildResp.json() as Promise<BuildDetail>)
                        .then((buildDetail) => {
                          if (buildDetail.timestamp) {
                            res(buildDetail.timestamp);
                          } else {
                            rej("No timestamp found");
                          }
                        })
                        .catch(rej);
                    })
                    .catch(rej);
                } else {
                  rej("No timestamp found");
                }
              } else {
                rej("No lastFailedBuild found");
              }
            })
            .catch(rej);
        })
        .catch(rej);
    });
  }
}

File: src/main/kotlin/io/frankmayer/dwf/endpoint/JenkinsEndpoint.kt
package io.frankmayer.dwf.endpoint

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import io.frankmayer.dwf.api.jenkins.Job
import io.frankmayer.dwf.config.EndpointConfig
import io.frankmayer.dwf.lib.Result
import java.time.Duration
import java.util.*

class JenkinsEndpoint(config: EndpointConfig) : Endpoint(config) {
    private val jenkinsToken = env("TOKEN")
    private val jenkinsUser = env("USER")
    private var token: String? = if (jenkinsUser != null && jenkinsToken != null) {
        Base64.getEncoder().encodeToString(("$jenkinsUser:$jenkinsToken").toByteArray())
    } else {
        null
    }
    private val objectMapper = ObjectMapper().configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)

    override fun request(project: String, workflow: String?): Result<String, String> {
        val reqProp = hashMapOf<String, String>()
        if (token != null) {
            reqProp["Authorization"] = "Basic $token"
        }

        return fetch(
            "${config.url}/job/$project/api/json?pretty=false&tree=lastBuild[timestamp],lastUnsuccessfulBuild[timestamp],lastFailedBuild[timestamp]",
            reqProp
        ).map({
            val job = objectMapper.readValue(it, Job::class.java)

            if (job.lastBuild == null) {
                Result.success(humanReadable(Duration.ZERO))
            } else {
                val lastFailed = job.lastFailedBuild ?: job.lastUnsuccessfulBuild

                if (lastFailed == null) {
                    Result.success(neverFailed)
                } else if (job.lastBuild.timestamp < lastFailed.timestamp) {
                    Result.success(humanReadable(Duration.ZERO))
                } else {
                    Result.success(
                        humanReadable(
                            Duration.ofMillis(
                                System.currentTimeMillis() - lastFailed.timestamp
                            )
                        )
                    )
                }
            }
        },
            {
                Result.failure("Jenkins responded with $it")
            })
    }
}
